volta: init at 1.1.1

cideM opened the pull request:
###### Description of changes

Volta is a newer Node version manager.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cideM committed:
volta: init at 1.1.1

Co-authored-by: Jules Amonith <examosa@fastmail.com>
Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

nicknovitski commented:
What's the behavior of volta when the `VOLTA_HOME` environment variable is unset?  Or more generally, after installing this package, what are the steps someone would have to do to use it?  Even if a wrapper set `VOLTA_HOME` to a mutable directory which volta then put its shims into, how would people add those shims to their `PATH`?

Maybe those questions don't need to be answered before this package is included, though. 😄

cideM commented:
Sorry for the long radio silence, I sort of forgot about this PR.

[This page](https://docs.volta.sh/advanced/installers#skipping-volta-setup) explains the setup process quite nicely. Specifically (emphasis mine):

> The Volta shim directory needs to be added to the PATH as well, so that the shims will work as expected as well. The shim directory is at $VOLTA_HOME/bin (%VOLTA_HOME%\bin on Windows), **where VOLTA_HOME is defaulted to:**
> 
>     ~/.volta on Unix
>     %LOCALAPPDATA%\Volta on Windows
> 

The Nix package will only handle the first mandatory step, that is, distributing the binaries. The shell setup can be done either manually, by updating `PATH`, or through `volta setup`

nicknovitski commented:
Cool!  I can confirm this works on aarch64-darwin:

```
% echo $VOLTA_HOME

% nix run github:cideM/nixpkgs/add-volta#volta -- install node@lts
success: installed and set node@18.15.0 (with npm@9.5.0) as default
% nix run github:cideM/nixpkgs/add-volta#volta -- which node
/Users/nick/.volta/tools/image/node/18.15.0/bin/node
```

The package looks good to me as-is, although one thing I can suggest is including shell completions with `volta completion [shell]` and installShellFiles (https://nixos.org/manual/nixpkgs/stable/#installshellfiles).

cideM commented:
Thanks for the feedback! I'd really like to add the completion step then, it's a very good idea.

examosa commented on pkgs/development/tools/volta/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+  cargoHash = "sha256-g6Ai0QN9cDiNX/YHdMM7/llJ/pCWZtv66wTaKvuN+bE=";
+
+  buildInputs = lib.optionals stdenv.isDarwin [darwin.apple_sdk.frameworks.Security libiconv];
+
```
In case you need any help with the completions, I had been working on a derivation for this as well:

```suggestion
  buildInputs = [installShellFiles]
  ++ lib.optionals stdenv.isDarwin [darwin.apple_sdk.frameworks.Security libiconv];

  postInstall = ''
    installShellCompletion --cmd volta \
      --bash <($out/bin/volta completions bash) \
      --fish <($out/bin/volta completions fish) \
      --zsh <($out/bin/volta completions zsh)
  '';
```

cideM commented:
I committed your change. Now I'm investigating an error about `homeless-shelter` in the `cargoInstallHook`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/968

SuperSandro2000 commented on pkgs/development/tools/volta/default.nix:
```diff
@@ -0,0 +1,48 @@
+    '';
+    homepage = "https://volta.sh/";
+    changelog = "https://github.com/volta-cli/volta/blob/main/RELEASES.md";
+    license = with licenses; [bsd2];
+    maintainers = with maintainers; [fbrs];
```
```suggestion
    license = with licenses; [ bsd2 ];
    maintainers = with maintainers; [ fbrs ];
```
don't use alejandra. It is very inconsistent with itself and adds a lot of extra space in the beginning of the file while clumsing things here together.

cideM replied:
I like that it's so opinionated but I'll make sure to not use it in `nixpkgs` then

SuperSandro2000 commented on pkgs/development/tools/volta/default.nix:
```diff
@@ -0,0 +1,48 @@
+
+  cargoHash = "sha256-g6Ai0QN9cDiNX/YHdMM7/llJ/pCWZtv66wTaKvuN+bE=";
+
+  buildInputs = [installShellFiles]
+  ++ lib.optionals stdenv.isDarwin [darwin.apple_sdk.frameworks.Security libiconv];
```
```suggestion
  buildInputs = [ installShellFiles ]
    ++ lib.optionals stdenv.isDarwin [ darwin.apple_sdk.frameworks.Security libiconv ];
```

cideM commented:
@SuperSandro2000 I committed your changes and squashed everything into one commit

drupol reviewed (changes requested):
Eval is broken, can you please fix the Cargo stuff ?

```
> source = "git+https://github.com/mikrostew/semver?branch=new-parser#7583eb352dc181ccd09978fd2b16461c1b1669c1"
       > source = "git+https://github.com/mikrostew/semver-parser?branch=rewrite#f5c74268a09eef16a289a667ca7b4925e690fe13"
       >
       > ERROR: The Cargo.lock contains git dependencies
       >
       > This is currently not supported in the fixed-output derivation fetcher.
       > Use cargoLock.lockFile / importCargoLock instead.```

cideM commented:
I'll fix it, although this is strange. I can build the package just fine and when I ran `nixpkgs-review` I now got an error but only the first time. If I re-run it it works just fine. I guess I had some incorrect assumptions about this command.

cideM commented:
@drupol I added the lock file and the three missing hashes for the Git dependencies. I can build just fine with `nix-build -A volta .` and I ran `nixpkgs-review rev` with my local commit after a `nix-store --gc`.

cideM commented:
I'm a little unsure if I still `cargoHash`. My thinking is that now that I'm vendoring dependencies with `Cargo.lock` there shouldn't be a need to supply the `cargoHash` field any more. I removed it and ran `nixpkgs-review` and it seems to work without it as well.

I checked the source for `pkgs/build-support/rust/build-rust-package/default.nix` and it seems that `cargoHash` won't be necessary when using `importCargoLock`. I removed it.

drupol commented on pkgs/development/tools/volta/default.nix:
```diff
@@ -0,0 +1,55 @@
+
+  src = fetchFromGitHub {
+    owner = "volta-cli";
+    repo = pname;
```
Use the name of the repo instead of a variable.

cideM replied:
Done
